lib/mp4/transmuxer.js

```diff
@@ -291,7 +381,197 @@ VideoSegmentStream = function(track) {
+
+    for (i = 0; i < nalUnits.length; i++) {
+      currentNal = nalUnits[i];
+
+      if (currentNal.nalUnitType === 'access_unit_delimiter_rbsp') {
+        if (currentFrame && currentFrame.length) {
+          frames.push(currentFrame);
+        }
+
+        currentFrame = [currentNal];
+      } else {
+        currentFrame.push(currentNal);
+      }
+    }
+    // push the final frame
+    frames.push(currentFrame);
+    return frames;
+  };
+
+  // convert an array of frames into an array of Gop with each Gop being
+  // composed of the frames that make up that Gop
+  this.groupFramesIntoGops_ = function(frames) {
+    var
+      i,
+      currentFrame,
+      currentGop = [],
+      gops = [];
+
+    for (i = 0; i < frames.length; i++) {
+      currentFrame = frames[i];
+
+      if (this.frameIsIFrame_(currentFrame)) {
+        if (currentGop && currentGop.length) {
+          gops.push(currentGop);
+        }
+
+        currentGop = [currentFrame];
+      } else {
+        currentGop.push(currentFrame);
+      }
+    }
+    // push the final Gop
+    gops.push(currentGop);
+    return gops;
+  };
+
+  // generate the track's sample table from an array of frames
+  this.generateSampleTable_ = function(frames, baseDataOffset) {
+    var
+      i,
+      sample,
+      currentFrame,
+      nextFrame,
+      firstNal,
+      lastNal,
+      frameDataSize,
+      currentSample,
+      dataOffset = baseDataOffset || 0,
+      samples = [];
+
+    for (i = 0; i < frames.length; i++) {
+      currentFrame = frames[i];
+      nextFrame = frames[i + 1];
+
+      firstNal = currentFrame[0];
+      lastNal = currentFrame[currentFrame.length - 1];
+      frameDataSize = sumByteLengths(currentFrame, 'data');
+
+      sample = this.createDefaultSample_();
+      sample.dataOffset = dataOffset;
+      sample.compositionTimeOffset = firstNal.pts - firstNal.dts;
+
+      if (nextFrame) {
+        sample.duration = nextFrame[0].dts - firstNal.dts;
+      } else {
+        sample.duration = lastNal.dts - firstNal.dts;
+
+        if (sample.duration === 0 &&
+            samples.length > 0) {
+          // for the last frame, copy the duration of the previous
+          // frame
+          sample.duration = samples[samples.length - 1].duration;
+        }
+      }
+
+      sample.size = 4 * currentFrame.length; // Space for nal unit size
+      sample.size += frameDataSize;
+
+      if (this.frameIsIFrame_(currentFrame)) {
+        sample.flags.dependsOn = 2;
+      }
+
+      dataOffset += sample.size;
+
+      samples.push(sample);
+    }
+    return samples;
+  };
+
+  // generate the track's raw mdat data from an array of frames
+  this.concatenateNalData_ = function (frames) {
+    var
+      i, j,
+      currentFrame,
+      currentNal,
+      dataOffset = 0,
+      nalsByteLength = frames.reduce(function(v, frame) {return v + sumByteLengths(frame, 'data'); }, 0),
+      numberOfNals = frames.reduce(function(v, frame) { return v + frame.length; }, 0),
```

Comment:
These two reduces() seem like more points in favor of building up and storing information about frames, rather than recalculate it over and over.